### PR TITLE
Expands codegenDecl to work in function params. fixes #22306

### LINKED ISSUE #22306: Expands `codegenDecl` to work in function params.
### Summary

By allowing the user to specify `condegenDecl` the interop can be improved further, we could do for instance:

```nim
{.emit:"""/*TYPESECTION*/
  int operate(int x, int y, int (*func)(const int&, const int&)){
    return func(x, y);
  };
""".}

proc operate(x, y: int32, fn: proc(x, y: int32 ): int32 {.cdecl.}): int32 {.importcpp:"$1(@)".}
proc add(a {.codegenDecl:"const $#& $#".}, b {.codegenDecl:"const $# $#", byref.}: int32): int32  {.cdecl.} = a + b
echo operate(1, 2, add)
``` 

### Description

See the code sample above

### Alternatives

_No response_

### Examples

_No response_

### Backwards Compatibility

_No response_

### Links

_No response_

### PR DESCRIPTION
**Pull request by jmgomez**

### CODE REVIEW

**Comment by juancarlospaco** [comment]
Very cool!. This works on C too ?, should be tested too.

**Comment by jmgomez** [comment]
> Very cool!. 

Thanks! The only thing left is to convince @Araq about `std/cpputils.nim`, so we can finally have built in`cppconst` (and others) without adding yet another pragma :P 

> This works on C too ?, should be tested too.

Yes, and there is nothing specific about C++ in the code so it would be redundant. The only reason why I added a `cpp` test is because none of the `codegenDecl` is tested for the c/cpp backend. I plan to add tests to it in the near future to showcase the other uses of it. In general it makes more sense to use it in Cpp as it would serve as an example of usage too.

**Comment by juancarlospaco** [comment]
Convincing Araq is the hardest part of Nim dev. 🤪

I can imagine uses of that new codegenDecl for JS backend in the future.

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -2432,7 +2432,7 @@ proc matchesAux(c: PContext, n, nOrig: PNode, m: var TCandidate, marker: var Int
     return
 
   template checkConstraint(n: untyped) {.dirty.} =
-    if not formal.constraint.isNil:
+    if not formal.constraint.isNil and not (formal.constraint.kind == nkStrLit and "$#" in formal.constraint.strVal):
```
Comment (Araq) [inline compiler/sigmatch.nim]: No. Too hacky.

Comment (jmgomez) [reply compiler/sigmatch.nim]: I thought it but didnt want to introduce "yet another flag", I assumed you had your reason when you did/accepted `codegenDecl` in the first place. 
But rethinking it now, maybe we could have `sfCodegenDecl` and that would add clarity to all checks over the place? (right now they mostly do `constraint != nil`)

Comment (Araq) [reply compiler/sigmatch.nim]: `sfCodegenDecl` is better.

**Pull request merged**